File: daily/112-Path-sum.py
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        if not root:
            return False
        def dfs(node, currentSum):
            if not node:
                return False
            currentSum += node.val
            if not node.left and not node.right:
                return currentSum == targetSum

            left = dfs(node.left, currentSum)
            right = dfs(node.right, currentSum)
             
            return left or right
        return dfs(root, 0)

File: daily/Binary Tree/112-Path-sum.go
package daily

func hasPathSum(root *TreeNode, targetSum int) bool {
	if root == nil {
		return false
	}

	var dfs func(node *TreeNode, currentSum int) bool
	dfs = func(node *TreeNode, currentSum int) bool {
		if node == nil {
			return false
		}

		currentSum += node.Val
		if node.Left == nil && node.Right == nil {
			return currentSum == targetSum
		}
		left := dfs(node.Left, currentSum)
		right := dfs(node.Right, currentSum)
		return left || right
	}
	return dfs(root, 0)
}
